lib/active_merchant/billing/gateways/payflow.rb

```diff
@@ -188,6 +189,53 @@ def build_credit_card_request(action, money, credit_card, options)
           end
         end
         xml.target!
+        add_level_two_three_fields(xml, options)
+      end
+
+      def add_level_two_three_fields(xml, options)
+        if options[:level_two_fields] || options[:level_three_fields]
+          xml_doc = Nokogiri::XML.parse(xml.target!) ## so it doesn't put on an xml instruction
+          %i[level_two_fields level_three_fields].each do |fields|
+            xml = add_fields(xml_doc, options[fields]) if options[fields]
+          end
+        end
+        xml
+      end
+
+      def check_fields(parent, fields, xml_doc)
+        fields.each do |k, v|
+          if v.is_a? String
+            new_node = Nokogiri::XML::Node.new(k, xml_doc)
+            new_node.add_child(v)
+            xml_doc.at_css(parent).add_child(new_node)
+          else
+            check_subparent_before_continuing(parent, k, xml_doc)
+            check_fields(k, v, xml_doc)
+          end
+        end
+        xml_doc
+      end
+
+      def check_subparent_before_continuing(parent, subparent, xml_doc)
+        unless xml_doc.at_css(subparent)
+          subparent_node = Nokogiri::XML::Node.new(subparent, xml_doc)
+          xml_doc.at_css(parent).add_child(subparent_node)
+        end
+      end
+
+      def string_on_fields(xml_doc, fields)
+        check_fields('Invoice', fields, xml_doc)
+        xml_doc.root.to_s
+      end
+
+      def add_fields(xml_doc, options_fields)
+        if options_fields.is_a? Hash
+          fields = options_fields
+        elsif options_fields.is_a? String
+          fields = JSON.parse(options_fields)
+        end
```

Comment:
Do we support the l2 and l3 fields being passed either as a Ruby Hash or a JSON string? I saw the latter in the remote tests, but not the former. I think maybe you said you were going to add some more tests; wouldn't hurt to add a test or two exercising the Hash path.

Reply:
Having messed around with the tests a bit, I think it is most straightforward to insist on them being a string. Have cleaned this up by getting rid of the ambiguously-named `string_on_fields` method.